Add client cert

siyengar opened the pull request:
Add methods needed for client certificate authentication support.

Commit:
Add options for client certificate authentication.

This adds options to the nifty server to support client
cert authentication.

Commit:
Add helpers to get SSL properties

Adds helper functions to get SSL properties from
the openssl engine. This enables features like authorization
of client certificates.

haijunz commented on nifty-core/src/main/java/com/facebook/nifty/core/NettyServerTransport.java:
```diff
@@ -131,6 +131,7 @@ public ChannelPipeline getPipeline()
                 SslServerConfiguration serverConfiguration = sslConfiguration.get();
                 if (serverConfiguration != null) {
                     SslHandler handler = serverConfiguration.createHandler();
+                    handler.setEnableRenegotiation(false);
```
Why disable this and how this is related to client cert support?

siyengar replied:
ya i should remove this from this particular diff. This is not needed for client support. I think this snuck in.

Commit:
Clean up constructor of NettyOpensslContext

There are too many params now. This cleans up the ctor.
In the next diff, the sslconfiguration will be needed for
the ssl session.

Commit:
Add SessionAwareSslHandler

This adds an SessionAwareSslHandler. Many components
might want to get access to an ssl session. The sslhandler
API is currently a bit racy, because calling it twice could
potentially cause renegotiation to trigger.

This changes it so that upstream filters will not get a
channelConnected until the SSL connection is established.

SessionAwareSslHandler propagates a message called sslsession
which upstream handlers can also register for.

Commit:
Add sslsession to connectioncontext

Adds sslsession to the nifty connection context.
This is useful because connection context is passed around.

haijunz commented:
LGTM
